Form Helper maxlength attribute uses validation #12773

lamasgergo opened the pull request:
Improvement for
#12773

Form helper now supports maxlength attribute from validation directly.

Rebased onto 3.next and squashed.

lamasgergo committed:
Form Helper maxlength attribute uses validation

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -1502,13 +1502,21 @@ protected function _magicOptions($fieldName, $options, $allowOverride)
             unset($options['step']);
         }
 
+        $validationLength = $context->getMaxLength($fieldName);
```
Shouldn't you first check if the field type is one of those in `$allowedTypes` below? If it's not then this method call is unnecessary overhead for other input types (e.g. `select`) which don't need `maxlength` attribute.

So I think this additional code should be move inside the `if` after `$allowedTypes` is initialized.

ADmad replied:
I understand that maxlength won't be set for e.g. for a textarea. But the `$context->getMaxLength()` call will still be made, which is unnecessary. That's my issue.

lamasgergo replied:
I think we have to rewrite this. Please check out my new file.
This will checking maxlength param too. Is this wrong? What you recommend?

ADmad commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -9574,4 +9576,114 @@ public function testControlLabelManipulationCheckboxes()
+     *
+     * @return void
+     */
+    public function testControlMaxLengthEntityContext()
```
There should also be a test showing that the maxlength set through validator takes precedence over maxlength deduced from table column length.

lamasgergo replied:
Ok i have added a test for it.

ADmad commented on src/View/Form/ContextInterface.php:
```diff
@@ -72,6 +72,14 @@ public function val($field);
+     * @param string $field A dot separated path to check required-ness for.
+     * @return int|null
+     */
+    public function getMaxLength($field);
```
We can't change the interface in minor release. Add a `@method` annotation instead at the top similar to what's done for `getRequiredMessage()`.

lamasgergo replied:
OK i have changed

ADmad commented on src/View/Form/EntityContext.php:
```diff
@@ -454,6 +454,29 @@ public function getRequiredMessage($field)
+        }
+        foreach ($validator->field($fieldName)->rules() as $rule) {
+            if ($rule->get('rule') === 'maxLength') {
+                return isset($rule->get('pass')[0]) ? $rule->get('pass')[0] : null;
```
Is the `isset()` check necessary?  You can't use max length validator without specifying the max length argument.

ADmad commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -9574,4 +9576,227 @@ public function testControlLabelManipulationCheckboxes()
+                'name' => 'title',
+                'type' => 'text',
+                'required' => 'required',
+                'maxlength' => 45,
```
Shouldn't this be `55` since that's what's set for maxLength validator? Validation rule length should take precedence over schema length.

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -1502,13 +1502,20 @@ protected function _magicOptions($fieldName, $options, $allowOverride)
-            && !empty($fieldDef['length'])
-            && $options['type'] !== 'select';
-
         $allowedTypes = ['text', 'textarea', 'email', 'tel', 'url', 'search'];
```
Nitpick: Can you please rename this variable to a more descriptive `$typesWithMaxlength`.

ADmad commented:
I wonder if all this extra code is really worth the overhead and extra maintenance it will need. Setting a maxLength validation with length smaller than table's column is pretty uncommon and for that edge case on can set `maxlength` in the FormHelper method options.

lamasgergo commented:
> I wonder if all this extra code is really worth the overhead and extra maintenance it will need. Setting a maxLength validation with length smaller than table's column is pretty uncommon and for that edge case on can set `maxlength` in the FormHelper method options.

I have started this pull request because i runned into in a project where i have a textarea and i have validated in model that, but the helper not generated maxlength attribute.
Ok i can add in the view  with `'maxlength' => 3000` but if i wrote the validation it would be great same magic as this reads out from schema i think.

You choose, close if it's not applicable.

ADmad commented on src/View/Form/FormContext.php:
```diff
@@ -167,6 +166,24 @@ public function getRequiredMessage($field)
+        }
+        foreach ($validator->field($field)->rules() as $rule) {
+            if ($rule->get('rule') === 'maxLength') {
+                return isset($rule->get('pass')[0])? $rule->get('pass')[0] : null;
```
The `isset()` check needs to be removed here too.

lamasgergo replied:
thanx,  i have removed

dereuromark commented:
I am just wondering what happened to the 100000 protection against upper values

    min($fieldDef['length'], 100000);

We dont need that anymore?
I feel like this change here could have been introduced in a less changing manner, as pure addition with falling back to the current behavior.

markstory commented on src/View/Form/ArrayContext.php:
```diff
@@ -229,6 +229,33 @@ public function getRequiredMessage($field)
+    /**
+     * Get field length from validation
+     *
+     * In this context class, this is simply defined by the 'maxlength' array.
```
```suggestion
     * In this context class, this is simply defined by the 'length' array.
```

lamasgergo replied:
thanx updated

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1502,13 +1502,20 @@ protected function _magicOptions($fieldName, $options, $allowOverride)
-            $options['maxlength'] = min($fieldDef['length'], 100000);
+        $typesWithMaxLength = ['text', 'textarea', 'email', 'tel', 'url', 'search'];
+        if (in_array($options['type'], $typesWithMaxLength)) {
+            $validationLength = $context->getMaxLength($fieldName);
```
Because this method isn't in the interface you should do a `method_exists()` check.

markstory replied:
This name could be `$maxLength` as it won't always be based on a validator.

lamasgergo replied:
thanx updated.

lamasgergo commented:
> I am just wondering what happened to the 100000 protection against upper values
> 
> ```
> min($fieldDef['length'], 100000);
> ```
> We dont need that anymore?
> I feel like this change here could have been introduced in a less changing manner, as pure addition with falling back to the current behavior.

I have added this protection. But why 100000? :) I was wondering on this line when i started this patch. But its ok users can't set maxlength more than 100000.

ADmad committed:
Simplify code.

Max length set in options should take highest precedence, followed by validator and lastly schema.

ADmad committed:
Ensure max lenght is non null.

ADmad commented:
The `if` block for deducing the max length was overly complicated so I have pushed some simplification.

The way it works now is:
- If `$options['maxlength']` do nothing more.
- Check if max length is available through validation rules and if it is use that.
- Check if max length is available through schema and if it is use that.

Max length set through options or validator *can* be larger than field length in schema. This is necessary as someone might be doing some pre-processing on the submitted value which could be longer than table's column length.
